app/src/main/kotlin/org/kotlinlsp/actions/Autocomplete.kt

```diff
@@ -1,34 +1,77 @@
 package org.kotlinlsp.actions
 
+import com.intellij.psi.util.leavesAroundOffset
+import com.intellij.psi.util.parentOfType
 import org.eclipse.lsp4j.CompletionItem
 import org.eclipse.lsp4j.CompletionItemKind
-import org.eclipse.lsp4j.InsertTextFormat
+import org.jetbrains.kotlin.analysis.api.KaExperimentalApi
+import org.jetbrains.kotlin.analysis.api.analyze
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtElement
 import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.psi.KtValueArgumentList
+import org.kotlinlsp.actions.completions.autoCompletionDotExpression
+import org.kotlinlsp.actions.completions.autoCompletionGeneric
 import org.kotlinlsp.index.Index
-import org.kotlinlsp.index.queries.getCompletions
+import org.kotlinlsp.index.db.Declaration
 
+
+@OptIn(KaExperimentalApi::class)
 fun autocompleteAction(ktFile: KtFile, offset: Int, index: Index): List<CompletionItem> {
-    val searchPrefix = computeSearchPrefix(ktFile, offset)
-    val completions = index.getCompletions(searchPrefix)
-        .map {
-            CompletionItem().apply {
-                label = it
-                kind = CompletionItemKind.Function
-                insertText = it
-                insertTextFormat = InsertTextFormat.PlainText
-            }
+    val leaf = ktFile.leavesAroundOffset(offset).asSequence()
+        .toList().last().first
+
+    val prefix = leaf.text.substring(0, offset - leaf.textRange.startOffset)
+    val completingElement = leaf.parentOfType<KtElement>() ?: ktFile
+
+    if (completingElement is KtNameReferenceExpression) {
+        if (completingElement.parent is KtDotQualifiedExpression) {
+            return completeDotQualified(ktFile, offset, index, completingElement.parent as KtDotQualifiedExpression, prefix)
+        } else {
+            return autoCompletionGeneric(ktFile, offset, index, completingElement, prefix)
```

Comment:
I'd try to refactor it so `autoCompletionGeneric` is called once, so the condition which triggers it is clearer